smack-extensions/src/main/java/org/jivesoftware/smackx/softwareinfo/form/SoftwareInfoForm.java

```diff
@@ -0,0 +1,272 @@
+import org.jivesoftware.smackx.xdata.packet.DataForm.Item;
+import org.jivesoftware.smackx.xdata.packet.DataForm.Type;
+
+/**
+ * {@link Form} that contains the software information.
+ * <br>
+ * Instance of {@link SoftwareInfoForm} can be created using {@link Builder#build()} method.
+ * <br>
+ * To obtain an instance of {@link Builder}, use {@link SoftwareInfoForm#getBuilder()} method.
+ * <br>
+ * An example to illustrate is provided inside SoftwareInfoFormTest inside the test package.
+ */
+public final class SoftwareInfoForm extends FilledForm {
+
+    public static final String FORM_TYPE = "urn:xmpp:dataforms:softwareinfo";
+
+    private SoftwareInfoForm(DataForm dataForm) {
+        super(dataForm);
+    }
+
+    /**
+     * Returns name of the OS used by client.
+     * <br>
+     * @return os
+     */
+    public String getOS() {
+        return readFirstValue("os");
+    }
+
+    /**
+     * Returns version of the OS used by client.
+     * <br>
+     * @return os_version
+     */
+    public String getOSVersion() {
+        return readFirstValue("os_version");
+    }
+
+    /**
+     * Returns name of the software used by client.
+     * <br>
+     * @return software
+     */
+    public String getSoftwareName() {
+        return readFirstValue("software");
+    }
+
+    /**
+     * Returns version of the software used by client.
+     * <br>
+     * @return software_version
+     */
+    public String getSoftwareVersion () {
+        return readFirstValue("software_version");
+    }
+
+    /**
+     * Returns the software icon if used by client.
+     * <br>
+     * @return {@link MediaElement} MediaElement or null
+     */
+    public MediaElement getIcon () {
+        FormField field = getField("icon");
+        if (field == null) {
+            return null;
+        }
+        FormFieldChildElement media = field.getFormFieldChildElement(MediaElement.QNAME);
+        if (media == null) {
+            return null;
+        }
+        return (MediaElement) media;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return EqualsUtil.equals(this, obj, (equalsBuilder, otherObj) -> {
+            equalsBuilder.append(getDataForm().getType(), otherObj.getDataForm().getType())
+                         .append(getDataForm().getTitle(), otherObj.getDataForm().getTitle())
+                         .append(getDataForm().getReportedData(), otherObj.getDataForm().getReportedData())
+                         .append(getDataForm().getItems(), otherObj.getDataForm().getItems())
+                         .append(getDataForm().getFields(), otherObj.getDataForm().getFields())
+                         .append(getDataForm().getExtensionElements(), otherObj.getDataForm().getExtensionElements());
+        });
+    }
+
+    @Override
+    public int hashCode() {
+        HashCode.Builder builder = HashCode.builder();
+        for (FormField field : getDataForm().getFields()) {
+            builder.append(field.getValues());
```

Comment:
This `getValues()` is wrong, as you do not compare the whole field.

You also may want to extend HashCode.Builder by `append(Collection<? extends Object> objects)`, akin to `append(Object[] array)`.